packages/backend-api/src/Controller/V1/Product/ProductController.php

```diff
@@ -0,0 +1,285 @@
+        $this->linksTransformer = $linksTransformer;
+        $this->availabilityFacade = $availabilityFacade;
+        $this->productDataFactory = $productDataFactory;
+        $this->productApiDataValidator = $productApiDataValidator;
+    }
+
+    /**
+     * Retrieves Product resource
+     * @Get("/products/{uuid}")
+     * @param string $uuid
+     * @return \Symfony\Component\HttpFoundation\Response
+     */
+    public function getProductAction(string $uuid): Response
+    {
+        $this->validateUuids([$uuid]);
+
+        $product = $this->productFacade->getByUuid($uuid);
+
+        $productArray = $this->productTransformer->transform($product);
+
+        $view = View::create($productArray, Response::HTTP_OK);
+
+        return $this->handleView($view);
+    }
+
+    /**
+     * Retrieves multiple Product resources
+     * @Get("/products")
+     * @QueryParam(name="page", requirements="-?\d+", default=1)
+     * @QueryParam(name="uuids", map=true, allowBlank=false)
+     * @param \FOS\RestBundle\Request\ParamFetcher $paramFetcher
+     * @param \Symfony\Component\HttpFoundation\Request $request
+     * @return \Symfony\Component\HttpFoundation\Response
+     */
+    public function getProductsAction(ParamFetcher $paramFetcher, Request $request): Response
+    {
+        $page = (int)$paramFetcher->get('page');
+
+        $query = new ProductQueryParams($this->pageSize, $page);
+
+        $filterUuids = $paramFetcher->get('uuids');
+        if (is_array($filterUuids)) {
+            $this->validateUuids($filterUuids);
+
+            $query = $query->withUuids($filterUuids);
+        }
+
+        $productsResult = $this->productFacade->findByProductQueryParams($query);
+
+        if ($page > $productsResult->getPage() || $page < 0) {
+            throw new UnprocessableEntityHttpException('There are no products on provided page.');
+        }
+
+        $productsArray = array_map(function (Product $product) {
+            return $this->productTransformer->transform($product);
+        }, $productsResult->getResults());
+
+        $links = $this->linksTransformer->fromPaginationResult($productsResult, $request->getUri());
+
+        $view = View::create($productsArray, Response::HTTP_OK, ['Link' => $links->format()]);
+
+        return $this->handleView($view);
+    }
+
+    /**
+     * @Post("/products")
+     * @Put("/products/{uuid}")
+     * @param \Symfony\Component\HttpFoundation\Request $request
+     * @param string|null $uuid
+     * @return \Symfony\Component\HttpFoundation\Response
+     */
+    public function createProductAction(Request $request, ?string $uuid = null): Response
+    {
+        $errors = $this->validateCreateProductInputs($request, $uuid);
+
+        if (count($errors) > 0) {
+            return $this->handleView($this->createValidationView($errors));
+        }
+
+        $productData = $this->productDataFactory->createFromApi($request->request->all(), $uuid);
```

Comment:
I like how `$request->request->all()` is extracted into variable in `patchProductAction` method.
It's clear what this value represents. Can you do it here too, please?

Reply:
This validation method bothered me for a while, and even Tom wanted to keep it in this fashion I decided to rewrite it to validate only if the UUID is setted... so please check `validateCreatingProductWithDefinedUuid` and beginning of the `createProductAction` method changed to
```php
        if ($uuid) {
            $this->validateCreatingProductWithDefinedUuid($uuid);
        }

        $productApiData = $request->request->all();
        $errors = $this->productApiDataValidator->validateCreate($productApiData);
```

I hate passing optional parameters when the optionality can be solved in the caller, so this is the result I like